itou/scripts/management/commands/update_suspensions_end_at.py

```diff
@@ -0,0 +1,121 @@
+        colnames = ["approval_number", "suspension_start_at", "suspension_end_at"]
+        converters = {
+            "approval_number": str,
+            "suspensions_start_at": pandas.to_datetime,
+            "suspensions_end_at": pandas.to_datetime,
+        }
+        df = pandas.read_csv(csv_file, sep=" ", names=colnames, header=None, converters=converters)
+
+        unique_approval_numbers_from_file = set(df["approval_number"].to_list())
+        self.stdout.write(f"Total of unique approvals in input file: {len(unique_approval_numbers_from_file)}")
+
+        # Just for the records as we only have one.
+        # Should be treated manually.
+        excluded_suspensions = suspensions.exclude(approval__number__in=unique_approval_numbers_from_file)
+        for suspension in excluded_suspensions:
+            self.stdout.write(
+                "Suspension not in log file. "
+                f"Pk: {suspension.pk}, start_at: {suspension.start_at}, end_at: {suspension.end_at}"
+            )
+
+        suspensions = suspensions.filter(approval__number__in=unique_approval_numbers_from_file)
+        self.stdout.write(f"Problematic suspensions found in input file: {suspensions.count()}")
+
+        # Count duplicated lines.
+        # Generated dataframe:
+        #               approval_number  size
+        #        35509  9999922919054    4
+        df = df.groupby(by="approval_number", as_index=False).size().sort_values("size")
+
+        updated_suspensions_counter = 0
+        skipped_suspensions_no_duplicate_counter = 0
+        skipped_suspensions_other_reason_counter = 0
+        for _, row in df.iterrows():
+            # Only approvals with several suspensions have caused troubles.
+            if row["size"] == 1:
+                skipped_suspensions_no_duplicate_counter += 1
+                continue
+
+            try:
+                approval = Approval.objects.get(number=row["approval_number"])
+            except Approval.DoesNotExist:
+                self.stdout.write(f"Skipping approval not found: {row['approval_number']}")
+                skipped_suspensions_other_reason_counter += 1
+                continue
+
+            last_suspension = approval.suspension_set.order_by("end_at").last()
+            # Don't update suspensions updated by a user after the first script ran.
+            if last_suspension.updated_at and last_suspension.updated_at > FIRST_SCRIPT_RUNNING_DATE:
+                self.stdout.write(
+                    "Skipping suspension updated after script ran: "
+                    f"{last_suspension.pk}, start_at: {last_suspension.start_at}, end_at: {last_suspension.end_at}, "
+                    f"updated_at: {last_suspension.updated_at.date()}"
+                )
+                skipped_suspensions_other_reason_counter += 1
+                continue
+
+            last_suspension.end_at -= relativedelta(months=MONTHS_OF_PROLONGATION * (row["size"] - 1))
```

Comment:
J'suis toujours un peu réticent à faire de l'arithmétique plutôt que mettre une valeur absolue quand c'est possible, mais de ce que j'ai entendu de la discussion d'hier ça devrais pas poser problème dans le cas actuel.

Reply:
Hum, quelle valeur absolue voudrais-tu mettre ? Le nombre d'itérations est variable.

Reply:
J'avais en tête `last_suspension.end_at = last_suspension.start_at + relativedelta(months=Suspension.MAX_DURATION_MONTHS)` mais il y a peut-être des cas que je n'ai pas en tête ;).

Reply:
J'ai changé la logique à ce niveau là en reprenant la première mise à jour de date de fin.